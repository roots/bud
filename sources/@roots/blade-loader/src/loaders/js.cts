export default function (source: string) {
  return [...source.matchAll(/@js(?<content>[\s\S]*?)@endjs/g)]
    .map(match => match.groups.content)
    .filter(Boolean)
    .join(`\n`)
}
