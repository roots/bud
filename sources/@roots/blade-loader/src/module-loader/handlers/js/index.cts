import {extract} from '@roots/blade-loader/module-loader/handlers'

export const extension = `js`
export const pattern =
  /@module\(['"]js['"]\)(?<content>[\s\S]*?)@endmodule/g

export default (source: string) => extract(source, pattern)
