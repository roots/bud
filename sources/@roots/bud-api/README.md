<p align="center">
  <img src="https://cdn.roots.io/app/uploads/logo-bud.svg" height="100" alt="Bud" />
</p>

<p align="center">
  <img alt="MIT License" src="https://img.shields.io/github/license/roots/bud?color=%23525ddc&style=flat-square" /> <img alt="npm" src="https://img.shields.io/npm/v/@roots/bud.svg?color=%23525ddc&style=flat-square" /> <img alt="Follow Roots" src="https://img.shields.io/twitter/follow/rootswp.svg?color=%23525ddc&style=flat-square" />
</p>

<h1 align="center">
  <strong>@roots/bud-api</strong>
</h1>

<p align="center">
  Bud configuration API
</p>

## Installation

Install **@roots/bud-api** to your project.

```shell
yarn add @roots/bud-api --dev
```

## Documentation

For more information on utilizing this package [check out our dedicated docs](https://bud.js.org)

## Community

Keep track of development and community news.

- Participate on the [Roots Discourse](https://discourse.roots.io)
- Follow [@rootswp on Twitter](https://twitter.com/rootswp)
- Read and subscribe to the [Roots blog](https://roots.io/blog/)
- Subscribe to the [Roots Newsletter](https://roots.io/subscribe/)
- Listen to the [Roots Radio podcast](https://roots.io/podcast/)

## Contributing

Contributions are welcome from everyone.

We have [contribution guidelines](https://github.com/roots/guidelines/blob/master/CONTRIBUTING.md) to help you get started.

## Sponsors

Help support our open-source development efforts by [becoming a patron](https://www.patreon.com/rootsdev).

<a href="https://k-m.com/"><img src="https://cdn.roots.io/app/uploads/km-digital.svg" alt="KM Digital" width="200" height="150"/></a>
<a href="https://carrot.com/"><img src="https://cdn.roots.io/app/uploads/carrot.svg" alt="Carrot" width="200" height="150"/></a>
<a href="https://www.c21redwood.com/"><img src="https://cdn.roots.io/app/uploads/c21redwood.svg" alt="C21 Redwood Realty" width="200" height="150"/></a>
<a href="https://wordpress.com/"><img src="https://cdn.roots.io/app/uploads/wordpress.svg" alt="WordPress.com" width="200" height="150"/></a>
<a href="https://pantheon.io/"><img src="https://cdn.roots.io/app/uploads/pantheon.svg" alt="Pantheon" width="200" height="150"/></a>
