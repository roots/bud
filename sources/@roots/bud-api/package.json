{
  "name": "@roots/bud-api",
  "version": "0.0.0",
  "description": "bud.js core module",
  "engines": {
    "node": ">=16"
  },
  "contributors": [
    {
      "name": "Kelly Mears",
      "email": "developers@tinypixel.dev",
      "url": "https://github.com/kellymears"
    },
    {
      "name": "Ben Word",
      "email": "ben@benword.com",
      "url": "https://github.com/retlehs"
    },
    {
      "name": "QWp6t",
      "url": "https://github.com/QWp6t"
    },
    {
      "name": "Brandon",
      "email": "brandon@tendency.me",
      "url": "https://github.com/Log1x"
    }
  ],
  "license": "MIT",
  "homepage": "https://roots.io/bud",
  "funding": {
    "type": "github sponsors",
    "url": "https://github.com/sponsors/roots"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/roots/bud.git",
    "directory": "sources/@roots/bud-api"
  },
  "bugs": "https://github.com/roots/bud/issues",
  "keywords": [
    "bud",
    "bud-framework"
  ],
  "files": [
    "lib",
    "src",
    "vendor"
  ],
  "type": "module",
  "exports": {
    ".": "./lib/index.js",
    "./api": "./lib/api/index.js",
    "./methods": "./lib/methods/index.js",
    "./methods/*": "./lib/methods/*.js"
  },
  "typesVersions": {
    "*": {
      ".": [
        "./lib/index.d.ts"
      ],
      "api": [
        "./lib/api/index.d.ts"
      ],
      "methods": [
        "./lib/methods/index.d.ts"
      ],
      "methods/*": [
        "./lib/methods/*.d.ts"
      ]
    }
  },
  "types": "./lib/index.d.ts",
  "module": "./lib/index.js",
  "devDependencies": {
    "@roots/bud-minify": "workspace:*",
    "@skypack/package-check": "0.2.2",
    "@types/node": "18.16.16"
  },
  "dependencies": {
    "@roots/bud-extensions": "workspace:*",
    "@roots/bud-framework": "workspace:*",
    "@roots/bud-support": "workspace:*",
    "@roots/container": "workspace:*",
    "chalk": "5.2.0",
    "tslib": "2.5.3"
  },
  "volta": {
    "extends": "../../../package.json"
  }
}
