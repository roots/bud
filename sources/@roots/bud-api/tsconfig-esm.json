{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./lib/esm",
    "module": "esnext"
  },
  "references": [
    {
      "path": "./../bud-framework/tsconfig-esm.json"
    },
    {
      "path": "./../bud-support/tsconfig-esm.json"
    }
  ]
}
