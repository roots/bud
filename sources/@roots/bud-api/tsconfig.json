{
  "extends": "../../../config/tsconfig.json",
  "compilerOptions": {
    "paths": {
      "@roots/bud-api/*": ["./src/*"]
    },
    "strict": true,
    "rootDir": "./src",
    "outDir": "./lib",
  },
  "include": ["./src"],
  "exclude": ["./lib", "./node_modules", "./**/*.test.ts"],
  "references": [
    {"path": "../bud-build/tsconfig.json"},
    {"path": "../bud-extensions/tsconfig.json"},
    {"path": "../bud-framework/tsconfig.json"},
    {"path": "../bud-support/tsconfig.json"},
    {"path": "../bud-terser/tsconfig.json"},
    {"path": "../container/tsconfig.json"}
  ]
}
