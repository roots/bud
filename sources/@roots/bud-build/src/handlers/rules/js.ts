import type {Factory} from '../index.js'

export const js: Factory = ({filter, makeRule, path}) =>
  makeRule()
    .setTest(filter(`pattern.js`))
    .setInclude([() => path(`@src`)])
    .setUse([])
