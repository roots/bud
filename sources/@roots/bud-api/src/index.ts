// Copyright Â© Roots Software Foundation LLC
// Licensed under the MIT license.

/**
 * Extends bud.js with high-level facades to simplify common configuration tasks
 *
 * @see https://bud.js.org
 * @see https://github.com/roots/bud
 */

import type * as Alias from '@roots/bud-api/methods/alias'
import type * as Assets from '@roots/bud-api/methods/assets'
import type * as Bundle from '@roots/bud-api/methods/bundle'
import type * as CompilePaths from '@roots/bud-api/methods/compilePaths'
import type * as Config from '@roots/bud-api/methods/config'
import type * as CopyDir from '@roots/bud-api/methods/copyDir'
import type * as CopyFile from '@roots/bud-api/methods/copyFile'
import type * as Define from '@roots/bud-api/methods/define'
import type * as Devtool from '@roots/bud-api/methods/devtool'
import type * as Entry from '@roots/bud-api/methods/entry'
import type * as Experiments from '@roots/bud-api/methods/experiments'
import type * as Externals from '@roots/bud-api/methods/externals'
import type * as Hash from '@roots/bud-api/methods/hash'
import type * as Html from '@roots/bud-api/methods/html'
import type * as Minimize from '@roots/bud-api/methods/minimize'
import type * as Persist from '@roots/bud-api/methods/persist'
import type * as Provide from '@roots/bud-api/methods/provide'
import type * as Proxy from '@roots/bud-api/methods/proxy'
import type * as Runtime from '@roots/bud-api/methods/runtime'
import type * as Serve from '@roots/bud-api/methods/serve'
import type * as SetProxyUrl from '@roots/bud-api/methods/setProxyUrl'
import type * as SetPublicProxyUrl from '@roots/bud-api/methods/setPublicProxyUrl'
import type * as SetPublicUrl from '@roots/bud-api/methods/setPublicUrl'
import type * as SetUrl from '@roots/bud-api/methods/setUrl'
import type * as SplitChunks from '@roots/bud-api/methods/splitChunks'
import type * as Use from '@roots/bud-api/methods/use'
import type * as Watch from '@roots/bud-api/methods/watch'

import {default as Service} from '@roots/bud-api/service'

declare module '@roots/bud-framework' {
  interface Bud {
    /**
     * ## bud.alias
     *
     * {@link https://bud.js.org/docs/bud.alias ðŸ“• Documentation}
     */
    alias: (...params: Alias.Parameters) => Bud

    /**
     * ## bud.assets
     *
     * {@link https://bud.js.org/docs/bud.assets ðŸ“• Documentation}
     */
    assets: (...params: Assets.Parameters) => Bud

    /**
     * ## bud.bundle
     *
     * {@link https://bud.js.org/docs/bud.bundle ðŸ“• Documentation}
     */
    bundle(...params: Bundle.Parameters): Bud

    /**
     * ## bud.compilePaths
     *
     * {@link https://bud.js.org/docs/bud.compilePaths ðŸ“• Documentation}
     */
    compilePaths(...params: CompilePaths.Parameters): Bud

    /**
     * ## bud.config
     *
     * {@link https://bud.js.org/docs/bud.config ðŸ“• Documentation}
     */
    config(...params: Config.Parameters): Bud

    /**
     * ## bud.copy
     *
     * Alias for {@link Assets.assets bud.assets}
     *
     * {@link https://bud.js.org/docs/bud.assets ðŸ“• Documentation}
     */
    copy: (...params: Assets.Parameters) => Bud

    /**
     * ## bud.copyFile
     *
     * {@link https://bud.js.org/docs/bud.copyDir ðŸ“• Documentation}
     */
    copyDir(...params: CopyDir.Parameters): Bud

    /**
     * ## bud.copyFile
     *
     * {@link https://bud.js.org/docs/bud.copyFile ðŸ“• Documentation}
     */
    copyFile(...params: CopyFile.Parameters): Bud

    /**
     * ## bud.define
     *
     * {@link https://bud.js.org/docs/bud.define ðŸ“• Documentation}
     */
    define(...params: Define.Parameters): Bud

    /**
     * ## bud.devtool
     *
     * {@link https://bud.js.org/docs/bud.devtool ðŸ“• Documentation}
     */
    devtool(...params: Devtool.Parameters): Bud

    /**
     * ## bud.entry
     *
     * Indicate application entrypoint(s)
     *
     * {@link https://bud.js.org/docs/bud.entry ðŸ“• Documentation}
     *
     * @example
     * ```js
     * bud.entry('@src/index.js')
     * ```
     *
     * @example
     * ```js
     * bud.entry(['@src/index.js', '@src/another.js'])
     * ```
     *
     * @example
     * ```js
     * bud.entry('app', '@src/index.js')
     * ```
     *
     * @example
     * ```js
     * bud.entry({
     *   app: ['@src/index.js'],
     *   admin: ['@src/admin.js'],
     * })
     * ```
     */
    entry(...params: Entry.Parameters): Bud

    /**
     * ## bud.experiments
     *
     * {@link https://bud.js.org/docs/bud.experiments ðŸ“• Documentation}
     */
    experiments(...params: Experiments.Parameters): Bud

    /**
     * ## bud.externals
     *
     * {@link https://bud.js.org/docs/bud.externals ðŸ“• Documentation}
     */
    externals(...params: Externals.Parameters): Bud

    /**
     * ## bud.hash
     *
     * {@link https://bud.js.org/docs/bud.hash ðŸ“• Documentation}
     */
    hash(...params: Hash.Parameters): Bud

    /**
     * ## bud.html
     *
     * {@link https://bud.js.org/docs/bud.template ðŸ“• Documentation}
     */
    html(...params: Html.Parameters): Bud

    /**
     * @roots/bud-minify
     */
    minify: import('@roots/bud-minify').default

    /**
     * ## bud.minimize
     *
     * {@link https://bud.js.org/docs/bud.minimize ðŸ“• Documentation}
     */
    minimize(...params: Minimize.Parameters): Bud

    /**
     * ## bud.override
     *
     * Alias for {@link Config.config bud.config}
     *
     * {@link https://bud.js.org/docs/bud.config ðŸ“• Documentation}
     */
    override(...params: Config.Parameters): Bud

    /**
     * ## bud.persist
     *
     * {@link https://bud.js.org/docs/bud.persist ðŸ“• Documentation}
     */
    persist(...params: Persist.Parameters): Bud

    /**
     * ## bud.provide
     *
     * {@link https://bud.js.org/docs/bud.provide ðŸ“• Documentation}
     */
    provide(...params: Provide.Parameters): Bud

    /**
     * ## bud.proxy
     *
     * {@link https://bud.js.org/docs/bud.proxy ðŸ“• Documentation}
     */
    proxy(...params: Proxy.Parameters): Bud

    /**
     * ## bud.runtime
     *
     * {@link https://bud.js.org/docs/bud.runtime ðŸ“• Documentation}
     */
    runtime(...params: Runtime.Parameters): Bud

    /**
     * ## bud.serve
     *
     * {@link https://bud.js.org/docs/bud.serve ðŸ“• Documentation}
     */
    serve(...params: Serve.Parameters): Bud

    /**
     * ## bud.setProxyUrl
     *
     * {@link https://bud.js.org/docs/bud.setProxyUrl ðŸ“• Documentation}
     */
    setProxyUrl(...params: SetProxyUrl.Parameters): Bud

    /**
     * ## bud.setPublicProxyUrl
     *
     * {@link https://bud.js.org/docs/bud.SetPublicProxyUrl ðŸ“• Documentation}
     */
    setPublicProxyUrl(...params: SetPublicProxyUrl.Parameters): Bud

    /**
     * ## bud.setPublicUrl
     *
     * {@link https://bud.js.org/docs/bud.SetPublicUrl ðŸ“• Documentation}
     */
    setPublicUrl(...params: SetPublicUrl.Parameters): Bud

    /**
     * ## bud.setUrl
     *
     * {@link https://bud.js.org/docs/bud.setUrl ðŸ“• Documentation}
     */
    setUrl(...params: SetUrl.Parameters): Bud

    /**
     * ## bud.splitChunks
     *
     * {@link https://bud.js.org/docs/bud.splitChunks ðŸ“• Documentation}
     */
    splitChunks(...params: SplitChunks.Parameters): Bud

    /**
     * ## bud.template
     *
     * {@link https://bud.js.org/docs/bud.template ðŸ“• Documentation}
     *
     * @deprecated use {@link Bud.html bud.html}
     */
    template(...params: Html.Parameters): Bud

    /**
     * ## bud.use
     *
     * {@link https://bud.js.org/docs/bud.use ðŸ“• Documentation}
     */
    use(...params: Use.Parameters): Bud

    /**
     * ## bud.version
     *
     * {@link https://bud.js.org/docs/bud.version ðŸ“• Documentation}
     */
    version(...params: Hash.Parameters): Bud

    /**
     * ## bud.watch
     *
     * {@link https://bud.js.org/docs/bud.watch ðŸ“• Documentation}
     */
    watch(...params: Watch.Parameters): Bud

    /**
     * ## bud.config
     *
     * Alias for {@link Config.config bud.config}
     *
     * {@link https://bud.js.org/docs/bud.config ðŸ“• Documentation}
     */
    webpackConfig(...params: Config.Parameters): Bud
  }

  type Api = Service

  interface Services {
    api: Service
    extensions: import('@roots/bud-extensions').default
  }
}

export {Service as default}
