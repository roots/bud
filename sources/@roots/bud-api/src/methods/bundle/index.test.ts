import '@roots/bud-api'

import {factory} from '@repo/test-kit'
import {beforeEach, describe, expect, it} from 'vitest'

import {bundle} from './index.js'

describe(`bud.bundle`, () => {
  let bud
  let instance: typeof bundle

  beforeEach(async () => {
    bud = await factory()
    instance = bundle.bind(bud)
  })

  it(`should set the bundle using a string`, async () => {
    await instance(`react`).api.processQueue()

    expect(
      bud.hooks.filter(`build.optimization.splitChunks`),
    ).toMatchSnapshot()
  })

  it(`should set the bundle using a string name and a string test`, async () => {
    await instance(`react`, `react`).api.processQueue()

    expect(
      bud.hooks.filter(`build.optimization.splitChunks`),
    ).toMatchSnapshot()
  })

  it(`should set the bundle using a string name and regular expression test`, async () => {
    await instance(`react`, /react/).api.processQueue()

    expect(
      bud.hooks.filter(`build.optimization.splitChunks`),
    ).toMatchSnapshot()
  })

  it(`should set the bundle using a string name and array of strings test`, async () => {
    await instance(`react`, [`react`, `react-dom`]).api.processQueue()

    expect(
      bud.hooks.filter(`build.optimization.splitChunks`),
    ).toMatchSnapshot()
  })

  describe(`when hashing is enabled`, () => {
    it(`should include the hash in the \`filename\` property`, async () => {
      bud.context.hash = true
      await instance(`react`, [`react`, `react-dom`])

      expect(
        bud.hooks.filter(`build.optimization.splitChunks`).cacheGroups
          .react.filename,
      ).toMatchSnapshot(`js/bundle/react/[name].[contenthash].js`)
    })
  })
})
