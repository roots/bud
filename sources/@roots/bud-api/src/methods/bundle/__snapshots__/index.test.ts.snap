// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bud.bundle > should set the bundle using a string 1`] = `
{
  "automaticNameDelimiter": "/",
  "cacheGroups": {
    "react": {
      "filename": "js/bundle/react/[name].js",
      "idHint": "react",
      "priority": -10,
      "test": /\\[\\\\/\\]\\(react\\)\\[\\\\/\\]/,
    },
  },
  "chunks": "all",
  "minSize": 0,
}
`;

exports[`bud.bundle > should set the bundle using a string name and a string test 1`] = `
{
  "automaticNameDelimiter": "/",
  "cacheGroups": {
    "react": {
      "filename": "js/bundle/react/[name].js",
      "idHint": "react",
      "priority": -10,
      "test": /\\[\\\\/\\]\\(react\\)\\[\\\\/\\]/,
    },
  },
  "chunks": "all",
  "minSize": 0,
}
`;

exports[`bud.bundle > should set the bundle using a string name and array of strings test 1`] = `
{
  "automaticNameDelimiter": "/",
  "cacheGroups": {
    "react": {
      "filename": "js/bundle/react/[name].js",
      "idHint": "react",
      "priority": -10,
      "test": /\\[\\\\/\\]\\(react\\|react-dom\\)\\[\\\\/\\]/,
    },
  },
  "chunks": "all",
  "minSize": 0,
}
`;

exports[`bud.bundle > should set the bundle using a string name and regular expression test 1`] = `
{
  "automaticNameDelimiter": "/",
  "cacheGroups": {
    "react": {
      "filename": "js/bundle/react/[name].js",
      "idHint": "react",
      "priority": -10,
      "test": /react/,
    },
  },
  "chunks": "all",
  "minSize": 0,
}
`;

exports[`bud.bundle > when hashing is enabled > should include the hash in the \`filename\` property > js/bundle/react/[name].[contenthash].js 1`] = `"js/bundle/react/[name].[contenthash].js"`;
